### PR TITLE
MM-30048 added thread related socket messages

### PR DESCRIPTION
**Pull request by reflog**
#### Summary

Added 3 new socket events for thread update/read/follow

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-30048
#### Release Note

```release-note
New websocket events added: thread_updated,thread_follow_changed,thread_read_changed
```

### CODE REVIEW

File: app/notification.go [hunk 1]
```diff
@@ -160,24 +160,31 @@ func (a *App) SendNotifications(post *model.Post, team *model.Team, channel *mod
 	mentionedUsersList := make([]string, 0, len(mentions.Mentions))
 	updateMentionChans := []chan *model.AppError{}
 	mentionAutofollowChans := []chan *model.AppError{}
+	threadParticipants := []string{post.UserId}
+	if *a.Config().ServiceSettings.ThreadAutoFollow && post.RootId != "" {
+		if parentPostList != nil {
+			threadParticipants = append(threadParticipants, parentPostList.Posts[parentPostList.Order[0]].UserId)
```
Comment (jespino) [inline app/notification.go:166]: Are you fixing here the "autofollow" for the root post creator"?

Comment (reflog) [reply app/notification.go:166]: @jespino  yup! somehow missed it in the previous PR

File: store/sqlstore/post_store.go [hunk 2]
```diff
@@ -1976,7 +1976,7 @@ func (s *SqlPostStore) updateThreadsFromPosts(transaction *gorp.Transaction, pos
 		if thread, found := threadByRoot[rootId]; !found {
 			// calculate participants
 			var participants model.StringArray
-			if _, err := transaction.Select(&participants, "SELECT DISTINCT UserId FROM Posts WHERE RootId=:RootId", map[string]interface{}{"RootId": rootId}); err != nil {
+			if _, err := transaction.Select(&participants, "SELECT DISTINCT UserId FROM Posts WHERE RootId=:RootId OR Id=:RootId", map[string]interface{}{"RootId": rootId}); err != nil {
```
Comment (jespino) [inline store/sqlstore/post_store.go:1979]: Fixing something else here too, right? I think maybe this makes sense to be backported to 5.29

Comment (reflog) [reply store/sqlstore/post_store.go:1979]: @jespino  yup. same issue. the root posts author wasn't included in the participants

Comment (reflog) [reply store/sqlstore/post_store.go:1979]: once we are done with review, I'll create a backport PR for just these changes

File: api4/user_test.go [hunk 3]
```diff
@@ -5401,6 +5401,70 @@ func TestGetThreadsForUser(t *testing.T) {
...
+					require.Contains(t, thread.Participants, th.BasicUser.Id)
+					require.Contains(t, thread.Participants, th.BasicUser2.Id)
+				}
+			case <-time.After(3 * time.Second):
+				return
+			}
+		}
+	}()
+
+	require.Truef(t, userHasReceivedFollow, "User should have received %s event", model.WEBSOCKET_EVENT_THREAD_FOLLOW_CHANGED)
```
Comment (jespino) [inline api4/user_test.go:5463]: Can we make this more granular? something like "Call one thing, check the events sent by that things, reset the values, call another thing, check the events send by that second thing, etcetera"?

**Review by jespino — CHANGES_REQUESTED** [review]
Some small suggestions and questions, but in general looks good to me.

File: store/sqlstore/post_store.go (hunk 2 above)
Comment (jespino) [reply store/sqlstore/post_store.go:1979]: Great, that sounds good to me :)

File: api4/user_test.go (hunk 3 above)
Comment (reflog) [reply api4/user_test.go:5463]: @jespino - I guess, but... what for? :) It's a rather clear test the way it is right now, what would splitting into small subcases clarify?

Comment (jespino) [reply api4/user_test.go:5463]: Not necesarily clarifies, but avoids possible bugs generating events when are not expected, or not generating events on certain actions. It is not the same to say "This 3 actions emits this 3 events, than saying this action emit this event, this other action emits this other event and so on". This is not necessarily wrong, but is more error prone and can give more false negatives on the tests that a more specific approach.

Comment (reflog) [reply api4/user_test.go:5463]: Mkay. Noted :+1:

File: app/notification.go [hunk 4]
```diff
@@ -403,6 +410,21 @@ func (a *App) SendNotifications(post *model.Post, team *model.Team, channel *mod
 	}
 
 	a.Publish(message)
+	// If this is a reply in a thread, notify participants
+	if *a.Config().ServiceSettings.ThreadAutoFollow && post.RootId != "" {
```
Comment (ashishbhate) [inline app/notification.go:414]: Why only if `ThreadAutoFollow` is true? If this is false, and a user is following the thread, then will they not get a notification? 
And conversely, a user who participated in the thread but then unfollowed the thread will be notified?

Comment (reflog) [reply app/notification.go:414]: You are absolutely right @ashishbhate . I forgot to introduce the new CollapsedThreads config setting.

**Review by ashishbhate — APPROVED** [review]

**Comment by reflog** [comment]
@jespino - anything blocking this?

**Comment by jespino** [comment]
@reflog not really, only finding the time to review it properly.

**Review by jespino — APPROVED** [review]
LGTM

**Comment by reflog** [comment]
@ogi-m - some basic smoke testing around posts could be nice, just to check that I didn't break anything unexpected

**Review by ogi-m — APPROVED** [review]
Done smoke testing around posts, everything's working as expected. LGTM 👍

**Pull request merged**